arbor/simulation.cpp

```diff
@@ -209,173 +254,162 @@ void simulation_state::reset() {
 
     communicator_.reset();
 
-    local_spikes_->current().clear();
-    local_spikes_->previous().clear();
+    for (auto& spikes: local_spikes_) {
+        spikes.clear();
+    }
+
+    epoch_.reset();
 }
 
 time_type simulation_state::run(time_type tfinal, time_type dt) {
-    // Calculate the size of the largest possible time integration interval
-    // before communication of spikes is required.
-    // If spike exchange and cell update are serialized, this is the
-    // minimum delay of the network, however we use half this period
-    // to overlap communication and computation.
-    const time_type t_interval = min_delay_/2;
-
-    // task that updates cell state in parallel.
-    auto update_cells = [&] () {
+    // Progress simulation to time tfinal, through a series of integration epochs
+    // of length at most t_interval_. t_interval_ is chosen to be no more than
+    // than half the network minimum delay.
+    //
+    // There are three simulation tasks that can be run partially in parallel:
+    //
+    // 1. Update:
+    //    Ask each cell group to update their state to the end of the integration epoch.
+    //    Generated spikes are stored in local_spikes_->current().
+    //
+    // 2. Exchange:
+    //    Consume local spikes held in local_spikes_->previous(), and collect such spikes
+    //    from across all ranks.
+    //    Translate spikes to local postsynaptic spike events, to be appended to pending_events_.
+    //
+    // 3. Enqueue events:
+    //    Take events from pending_events_, together with any event-generator events for the
+    //    next epoch and any left over events from the last epoch, and collate them into
+    //    the per-cell event_lanes for the next epoch.
+    //
+    // Writing U(k) for Update on kth epoch; D(k) for Exchange of spikes generated in the kth epoch;
+    // and E(k) for Enqueue of the events required for the kth epoch, there are the following
+    // dependencies:
+    //
+    //     * E(k) precedes U(k).
+    //     * U(k) precedes D(k).
+    //     * U(k) precedes U(k+1).
+    //     * D(k) precedes E(k+2).
+    //     * D(k) precedes D(k+1).
+    //
+    // In the schedule implemented below, U(k) and D(k-1) or U(k) and E(k+1) can be run
+    // in parallel, while D and E operations must be serialized (D writes to pending_events_,
+    // while E consumes and clears it). The local spike collection and the per-cell event
+    // lanes are double buffered.
+    //
+    // Required state on run() invocation with epoch_.id==k:
+    //     * For k≥0,  U(k) and D(k) have completed.
```

Comment:
The way it has been rigged up, before and after calls to `run()`, `epoch_` will contain the most recently completed epoch, starting with the haven't-done-anytthing-yet-at-all epoch with id -1. In `run()`, the epoch `current` is first set to be the epoch following `epoch_`.

Reply:
Actually, having read that, it's probably sensible to define the epoch `prev` right at the start, too. I'll make a small edit.

Reply:
Okay, I was confused because the comment underneath lists the same requirements. 
```
// Requires state at end of run(), with epoch_.id==k:
//     * U(k) and D(k) have compelted.
```
But I guess that the `epoch` in this case has already been advanced.

Reply:
Yep, that was the intent. Would it aid clarity to use a different index than `k` here?

Reply:
I think it's okay :)